src/text/line.rs

```diff
@@ -586,31 +555,75 @@ impl WidgetRef for Line<'_> {
+                Some(Alignment::Center) => (area.width.saturating_sub(line_width)) / 2,
+                Some(Alignment::Right) => area.width.saturating_sub(line_width),
+                Some(Alignment::Left) | None => 0,
+            };
+            let area = area.indent_x(offset);
+            render_spans(&self.spans, area, buf, 0);
         } else {
             let offset = match self.alignment {
-                Some(Alignment::Center) => (area.width.saturating_sub(width)) / 2,
-                Some(Alignment::Right) => area.width.saturating_sub(width),
+                Some(Alignment::Center) => (line_width.saturating_sub(area.width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, offset);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_offset`, skip it
+///    2. If the span is partially before the `span_offset`, render the right side of the span
+///    3. If the span is completely after the `span_offset`, render the whole span
+/// 2. Indent `area` to account for the width of the rendered span
+/// 3. Update `span_offset` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(spans: &[Span], mut area: Rect, buf: &mut Buffer, mut span_offset: u16) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        if span_offset > 0 {
+            // the first visible span is only partially visible, so truncate the start and render it
+            // starting at span offset, all subsequent visible spans are rendered in full (as
+            // available space permits)
+            let available_width = span_width - span_offset;
+            let (content, actual_width) = span
+                .content
+                .unicode_truncate_start(available_width as usize);
```

Comment:
Why do you use `.unicode_truncate_start` here? The docs say:

> Truncates a string to be at most width in terms of display width by removing the start characters.

Shouldn't we use `.unicode_truncate` and `.unicode_truncate_start` depending on the `Line`'s alignment?

Reply:
The right truncation at end end happens when the area ends all by itself. That’s way Alignment::Left starts at 0 and at some point area.is_empty().

Center alignment is similar. Find the starting point and render until area.is_empty().

Basically: truncation on the right side happens due to the limited area. Truncation on the left is calculated and then done by the truncation crate.

Reply:
Clarified with a comment in c08cd0c4

Reply:
clarified with a comment in c08cd0c4